### ISSUE TITLE
react-redux: type error for connect inside HOC

### ISSUE THREAD

**Issue by OliverJAsh** [DefinitelyTyped/DefinitelyTyped#31363]
Using `connect` inside even the simplest of HOCs, using types v6.0.11, results in an unexpected type error.

**Update**: there is also a different but very similar issue which effects versions prior to v6.0.11. See https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31363#issuecomment-459303014.

Potentially related to https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30946 https://github.com/DefinitelyTyped/DefinitelyTyped/issues/29991

``` tsx
import { ComponentType, StatelessComponent } from 'react';
import { connect } from 'react-redux';

export const myHoc = function<ComposedComponentProps>(
    _ComposedComponent: ComponentType<ComposedComponentProps>,
) {
    declare const MyHOCComponent: StatelessComponent<ComposedComponentProps>;

    const enhance = connect();
    return enhance(
        /* Argument of type 'FunctionComponent<ComposedComponentProps>' is not assignable to parameter of type 'ComponentType<Matching<DispatchProp<AnyAction>, ComposedComponentProps>>'.
  Type 'FunctionComponent<ComposedComponentProps>' is not assignable to type 'FunctionComponent<Matching<DispatchProp<AnyAction>, ComposedComponentProps>>'.
    Types of property 'propTypes' are incompatible.
      Type 'WeakValidationMap<ComposedComponentProps> | undefined' is not assignable to type 'WeakValidationMap<Matching<DispatchProp<AnyAction>, ComposedComponentProps>> | undefined'.
        Type 'WeakValidationMap<ComposedComponentProps>' is not assignable to type 'WeakValidationMap<Matching<DispatchProp<AnyAction>, ComposedComponentProps>>'.
          Type '(null extends ComposedComponentProps[K] ? Validator<ComposedComponentProps[K] | null | undefined> : undefined extends ComposedComponentProps[K] ? Validator<ComposedComponentProps[K] | null | undefined> : Validator<...>) | undefined' is not assignable to type '(null extends Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] ? Validator<Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] | null | undefined> : undefined extends Matching<...>[K] ? Validator<...> : Validator<...>) | undefined'.
            Type 'null extends ComposedComponentProps[K] ? Validator<ComposedComponentProps[K] | null | undefined> : undefined extends ComposedComponentProps[K] ? Validator<ComposedComponentProps[K] | null | undefined> : Validator<...>' is not assignable to type '(null extends Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] ? Validator<Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] | null | undefined> : undefined extends Matching<...>[K] ? Validator<...> : Validator<...>) | undefined'.
              Type 'Validator<ComposedComponentProps[K] | null | undefined> | (undefined extends ComposedComponentProps[K] ? Validator<ComposedComponentProps[K] | null | undefined> : Validator<ComposedComponentProps[K]>)' is not assignable to type '(null extends Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] ? Validator<Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] | null | undefined> : undefined extends Matching<...>[K] ? Validator<...> : Validator<...>) | undefined'.
                Type 'Validator<ComposedComponentProps[K] | null | undefined>' is not assignable to type 'null extends Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] ? Validator<Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] | null | undefined> : undefined extends Matching<...>[K] ? Validator<...> : Validator<...>'.
                  Type 'null extends ComposedComponentProps[K] ? Validator<ComposedComponentProps[K] | null | undefined> : undefined extends ComposedComponentProps[K] ? Validator<ComposedComponentProps[K] | null | undefined> : Validator<...>' is not assignable to type 'null extends Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] ? Validator<Matching<DispatchProp<AnyAction>, ComposedComponentProps>[K] | null | undefined> : undefined extends Matching<...>[K] ? Validator<...> : Validator<...>'.ts(2345) */
        MyHOCComponent,
    );
};
```

---

If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/react-redux` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @tkqubo @kenzierocks @clayne11 @tansongyang @nicholasboll @mdibyo @pdeva @kallikrein @val1984 @jrakotoharisoa @apapirovski @surgeboris 

If you do not mention the authors the issue will be ignored.

**Comment by surgeboris** [comment]
What TS version are you using?

**Comment by OliverJAsh** [comment]
3.1.6

**Comment by surgeboris** [comment]
I've taken a look at your problem and it seems that I'm not yet ready to tackle it (feels like I don't have mental capacity to understand complex interaction of a very abstract types and inference).

The only thing I might suggest you - is to use explicit cast `as any as React.ComponentType<ResultingProps>` to the return value.

**Comment by OliverJAsh** [comment]
Do you know what change might have caused this? We had it working for
v6.0.8. If someone can tell me what changed since then, I'm happy to
investigate.

**Comment by surgeboris** [comment]
@OliverJAsh As far as I can see from the git history changes [changes in this commit are creating problems for you](https://github.com/surgeboris/DefinitelyTyped/commit/ee59ccab23114d85c2275194d9166f603c308153#diff-ee7ace1841a23c70c0c2b663cfae1cb7).

As a side note: you could have used [`git bisect`](https://git-scm.com/book/en/v2/Git-Tools-Debugging-with-Git#_binary_search) yourself to figure out the problematic commit.

**Comment by surgeboris** [comment]
I've tried to narrow down the scope of error, and it seems that the bug was introduced by [this type parameter `C` defined using `C` itself (recursively)](https://github.com/apapirovski/DefinitelyTyped/commit/dfbcd7c87cbb086640935d035ccbdf7efdf7a0f9#diff-ee7ace1841a23c70c0c2b663cfae1cb7R99).

Here's the reduced showcase that reproduces this issue's subject error with all excessive information removed:
```
class ReactComponent<TProps> {
    props: TProps;
}
function hoc<TComponentToWrapProps extends { user: number }>(arg: ReactComponent<TComponentToWrapProps>) {
    const enhance = connect(() => ({
        user: 1,
    }));
    return enhance(arg);
}
type GetProps<TReactComponent> = TReactComponent extends ReactComponent<infer TProps> ? TProps : never;
type Matching<InjectedProps, DecorationTargetProps> = {
    [Key in keyof InjectedProps]: Key extends keyof DecorationTargetProps ? DecorationTargetProps[Key] : InjectedProps[Key]
};
type MapStateToProps<TStateProps> = () => TStateProps;
function connect<TStateProps>(arg: MapStateToProps<TStateProps>) {
    return <C extends ReactComponent<Matching<TStateProps, GetProps<C>>>>(component: C) => {
        return null;
    };
}
```
If you replace `Matching<TStateProps, GetProps<C>>` with just `GetProps<C>` (or just `TStateProps`) the error will disappear.

I'm not sure if it is a type definition error or TS bug, as I do not understand neither how this recursively defined type parameter `C` is actually working nor how it's supposed to work.

@apapirovski It would be great if you could share your understanding of how this code works and what you tried to achieve with it...

Also, @Kovensky, sorry for bothering, but the author does not reply for 2 days (even 4 days, since initial highlight on issue opening) so there's a growing chance that he will not reply at all. You're seem to be smarter then me, so maybe you can wrap your hand around this obscure type? Or maybe you can advice someone else who can...

**Comment by clicktronix** [comment]
I have the same problem at #30946 , we thought that adding `Matching` in version 6.0.9 broke the types. Maybe this is the typescript problem. Matching looks right

**Comment by 19majkel94** [comment]
I have this error with `6.0.11`:
```
Type '{}' is missing the following properties from type 'Readonly<Pick<Props, "increment" | "incrementAsync">>': increment, incrementAsyncts(2739)
```

My connect code:
```ts
const mapState = (state: RootState): StateProps => ({
  value: state.counter.value,
});

const mapDispatch = (dispatch: Dispatch): DispatchProps => ({
  increment: dispatch.counter.increment,
  incrementAsync: dispatch.counter.incrementAsync,
});

export default connect(
  mapState,
  mapDispatch,
)(Counter);
```

**Comment by surgeboris** [comment]
@19majkel94
- I'm not sure why you're accessing `counter` property on dispatch in `mapDispatch` props, because `dispatch` is supposed to be a function in this case.
- I'm not able to reproduce your bug, as don't know the definitions of your `StateProps`, `DispatchProps`, `increment` and `incrementAsync`.
- Also, can you check if your code working with `6.0.9`?

**Comment by 19majkel94** [comment]
> I'm not sure why you're accessing counter property on dispatch in mapDispatch props, because dispatch is supposed to be a function in this case.

I am using [rematch](https://github.com/rematch/rematch), I forgot to mention that.

> I'm not able to reproduce your bug, as don't know the definitions of your StateProps, DispatchProps, increment and incrementAsync.

Here is a full code for the component:
```ts
import React, { FunctionComponent } from "react";
import { connect } from "react-redux";

import { RootState, Dispatch } from "@src/config/rematch";

interface StateProps {
  value: number;
}

interface DispatchProps {
  increment: () => void;
  incrementAsync: () => Promise<void>;
}

interface OwnProps {
  required: true;
}

type Props = StateProps & DispatchProps & OwnProps;

const Counter: FunctionComponent<Props> = ({
  value,
  increment,
  incrementAsync,
}) => (
  <>
    <h3>Counter value: {value}</h3>
    <button onClick={increment}>Increment</button>
    <button onClick={incrementAsync}>Increment after 1s</button>
  </>
);

const mapState = (state: RootState, ownProps: OwnProps): StateProps => ({
  value: state.counter.value,
});

const mapDispatch = (dispatch: Dispatch): DispatchProps => ({
  increment: dispatch.counter.increment,
  incrementAsync: dispatch.counter.incrementAsync,
});

export default connect(
  mapState,
  mapDispatch,
)(Counter);
```

And the model code:
```ts
import createModel from "@src/common/helpers/createModel";
import sleep from "@src/common/helpers/sleep";
import { dispatch } from "@src/config/rematch";

export interface CounterModelState {
  value: number;
}

const initialState: CounterModelState = {
  value: 0,
};

const counterModel = createModel<CounterModelState>()({
  name: "counter",
  state: initialState,
  reducers: {
    increment(state) {
      state.value++;
      return state;
    },
  },
  effects: {
    async incrementAsync() {
      await sleep(1000);
      dispatch.counter.increment();
    },
  },
});

export default counterModel;
```

> Also, can you check if your code working with 6.0.9?

Yes, I don't have this error anymore on `6.0.9` and other versions but instead I have this one:

![image](https://user-images.githubusercontent.com/10618781/50387120-3bb54380-06f5-11e9-9086-27f55f19dfad.png)

```
Argument of type '(dispatch: ExtractRematchDispatchersFromModels<{ counter: ModelWithState<CounterModelState, { name: string; state: CounterModelState; reducers: { increment(state: CounterModelState): CounterModelState; }; effects: { incrementAsync(this: { [key: string]: (payload?: any, meta?: any) => Action<...>; }): Promise<...>; }...' is not assignable to parameter of type 'MapDispatchToPropsParam<DispatchProps, OwnProps>'.
  Type '(dispatch: ExtractRematchDispatchersFromModels<{ counter: ModelWithState<CounterModelState, { name: string; state: CounterModelState; reducers: { increment(state: CounterModelState): CounterModelState; }; effects: { incrementAsync(this: { [key: string]: (payload?: any, meta?: any) => Action<...>; }): Promise<...>; }...' is not assignable to type 'MapDispatchToPropsFactory<DispatchProps, OwnProps>'.
    Types of parameters 'dispatch' and 'dispatch' are incompatible.
      Type 'Dispatch<Action<any>>' is missing the following properties from type 'ExtractRematchDispatchersFromModels<{ counter: ModelWithState<CounterModelState, { name: string; state: CounterModelState; reducers: { increment(state: CounterModelState): CounterModelState; }; effects: { incrementAsync(this: { [key: string]: (payload?: any, meta?: any) => Action<...>; }): Promise<...>; }; }>; route...': counter, router
```

**Comment by 19majkel94** [comment]
@surgeboris 
Ok, looks like 6.0.11 works ok if I turn off `strictFunctionTypes` option 😕 
```
  "strict": true,
  "strictFunctionTypes": false,
```
But I don't think it's React-Redux typings fault, as I'm expecting rich `Dispatch` object which is against TS `strictFunctionTypes` rule 😞

**Comment by clicktronix** [comment]
`"strictFunctionTypes": false` will not work for my case. @surgeboris @apapirovski The code crashes at the place of use of the `Matching` added in 6.0.9 #28189. It seems that the typescript do not work correctly, but I'm not completely sure. Here are two cases where matching works correctly and incorrectly:
```
type MyType = { a: string, b: boolean, c?: string };
type T = Matching<MyType, MyType>;
const o: T = null as any as MyType;           // Work correctly
```
Incorrect behavior:
```
return function component<TOwn>() {
...
type T = Matching<TOwn, TOwn>;
const o: T = null as any as TOwn;            // Error
}
```
Error description
```
Type 'TOwn' is not assignable to type 'Matching<TOwn, TOwn>'
```
But if you make the properties in the `Matching` optional, then everything works correctly
```
export type Matching<InjectedProps, DecorationTargetProps> = {
    [P in keyof DecorationTargetProps]?: P extends keyof InjectedProps
    .....
```

**Comment by surgeboris** [comment]
@clicktronix "strictFUnctionTypes" is not related to your problem, it's related to the usage of [`rematch` redux framework](https://github.com/rematch/rematch) which should be in a separate issue (but if author doesn't care about that - me neither).

If you make the properties in the `Matching` optional - that breaks tests and (from my point of view) kills the whole point of `Matching` type).

As I investigated earlier, your problem is that [`Matching` type is written in a recursive way](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/31363#issuecomment-448013978). I cannot do anything about it as I do not understand it completely.

**Comment by clicktronix** [comment]
@surgeboris Yes, I agree, the usefulness of the `Matching` will not be in this case.

**Comment by OliverJAsh** [comment]
> @apapirovski It would be great if you could share your understanding of how this code works and what you tried to achieve with it...
> 
> Also, @Kovensky, sorry for bothering, but the author does not reply for 2 days (even 4 days, since initial highlight on issue opening) so there's a growing chance that he will not reply at all. You're seem to be smarter then me, so maybe you can wrap your hand around this obscure type? Or maybe you can advice someone else who can...

I think we're still waiting on these replies?

**Comment by OliverJAsh** [comment]
The issue I described initially did not apply to 6.0.8 of the react-redux types. However, I'm now experiencing a bug when using `connect` inside a HOC with version 6.0.8 of the react-redux types.

This is probably a subtly different bug from the one I initially described at the top of this issue.

Note the code comment "error disappears when removing this constraint".

(I won't file a new issue for this since this bug is superseded by the one I described at the top of this issue, which applies to the latest version (6.0.11 at the time of writing).)

``` ts
import { ComponentType, StatelessComponent } from 'react';
import { connect, MapStateToPropsParam } from 'react-redux';
import { Dispatch } from 'redux';

type StateRoot = { foo: 1 };

// Default dispatch props for react-redux
type DefaultDispatchProps = { dispatch: Dispatch };

type RequiredProp = { foo: 1 };

// Error disappears when removing this constraint:
// export const myHoc = function<ComposedComponentProps>(
export const myHoc = function<ComposedComponentProps extends RequiredProp>(
    _ComposedComponent: ComponentType<ComposedComponentProps>,
) {
    type OwnProps = ComposedComponentProps;
    type StateProps = { user: 1 };
    type Props = DefaultDispatchProps & StateProps & OwnProps;

    declare const MyHOCComponent: StatelessComponent<Props>;

    const mapStateToProps: MapStateToPropsParam<
        StateProps,
        OwnProps,
        StateRoot
    > = () => ({
        user: 1,
    });

    const enhance = connect(mapStateToProps);
    return enhance(
        /* Argument of type 'FunctionComponent<Props>' is not assignable to parameter of type 'ComponentType<Shared<StateProps & DispatchProp<AnyAction>, Props>>'.
  Type 'FunctionComponent<Props>' is not assignable to type 'FunctionComponent<Shared<StateProps & DispatchProp<AnyAction>, Props>>'.
    Types of property 'propTypes' are incompatible.
      Type 'WeakValidationMap<Props> | undefined' is not assignable to type 'WeakValidationMap<Shared<StateProps & DispatchProp<AnyAction>, Props>> | undefined'.
        Type 'WeakValidationMap<Props>' is not assignable to type 'WeakValidationMap<Shared<StateProps & DispatchProp<AnyAction>, Props>>'.
          Type '(null extends Props[K] ? Validator<Props[K] | null | undefined> : undefined extends Props[K] ? Validator<Props[K] | null | undefined> : Validator<Props[K]>) | undefined' is not assignable to type '(null extends Shared<StateProps & DispatchProp<AnyAction>, Props>[K] ? Validator<Shared<StateProps & DispatchProp<AnyAction>, Props>[K] | null | undefined> : undefined extends Shared<...>[K] ? Validator<...> : Validator<...>) | undefined'.
            Type 'null extends Props[K] ? Validator<Props[K] | null | undefined> : undefined extends Props[K] ? Validator<Props[K] | null | undefined> : Validator<Props[K]>' is not assignable to type '(null extends Shared<StateProps & DispatchProp<AnyAction>, Props>[K] ? Validator<Shared<StateProps & DispatchProp<AnyAction>, Props>[K] | null | undefined> : undefined extends Shared<...>[K] ? Validator<...> : Validator<...>) | undefined'.
              Type 'Validator<Props[K] | null | undefined> | (undefined extends Props[K] ? Validator<Props[K] | null | undefined> : Validator<Props[K]>)' is not assignable to type '(null extends Shared<StateProps & DispatchProp<AnyAction>, Props>[K] ? Validator<Shared<StateProps & DispatchProp<AnyAction>, Props>[K] | null | undefined> : undefined extends Shared<...>[K] ? Validator<...> : Validator<...>) | undefined'.
                Type 'Validator<Props[K] | null | undefined>' is not assignable to type 'null extends Shared<StateProps & DispatchProp<AnyAction>, Props>[K] ? Validator<Shared<StateProps & DispatchProp<AnyAction>, Props>[K] | null | undefined> : undefined extends Shared<...>[K] ? Validator<...> : Validator<...>'.
                  Type 'null extends Props[K] ? Validator<Props[K] | null | undefined> : undefined extends Props[K] ? Validator<Props[K] | null | undefined> : Validator<Props[K]>' is not assignable to type 'null extends Shared<StateProps & DispatchProp<AnyAction>, Props>[K] ? Validator<Shared<StateProps & DispatchProp<AnyAction>, Props>[K] | null | undefined> : undefined extends Shared<...>[K] ? Validator<...> : Validator<...>'. */
        MyHOCComponent
    );
};
```

**Comment by Manc** [comment]
Admittedly I'm a bit confused with all this, but I think my problem may be related to this.

I use `"@types/react-redux": "^7.0.1"` and "suddenly" I get errors like this, where I use `connect`:

```
[tsl] ERROR in [...]/src/components/Header/index.tsx(161,41)
      TS2345: Argument of type 'typeof Header' is not assignable to parameter of type 'ComponentType<never>'.
  Type 'typeof Header' is not assignable to type 'ComponentClass<never, any>'.
    Types of property 'contextType' are incompatible.
      Type 'React.Context<any>' is not assignable to type 'import("[...]/node_modules/@types/react-redux/node_modules/@types/react/index").Context<any>'.
        Types of property 'Provider' are incompatible.
          Type 'React.ProviderExoticComponent<React.ProviderProps<any>>' is not assignable to type 'import("[...]/node_modules/@types/react-redux/node_modules/@types/react/index").ProviderExoticComponent<import("[...]/node_modules/@types/react-redux/node_modules/@types/react/index").ProviderProps<any>>'.
            Types of property 'propTypes' are incompatible.
              Type 'WeakValidationMap<ProviderProps<any>>' is not assignable to type 'ValidationMap<ProviderProps<any>>'.
                Property 'value' is optional in type 'WeakValidationMap<ProviderProps<any>>' but required in type 'ValidationMap<ProviderProps<any>>'.
```

I figured out (after a lot of searching), that this those errors are related to a package upgrade of `@types/react` version `16.7.14` or later. Looks like in that version `WeakValidationMap` was introduced (see https://github.com/DefinitelyTyped/DefinitelyTyped/commit/c2367fab52cf48fafa9daa915f0714a81d3c122f#diff-96b72df8b13a8a590e4f160cbc51f40c). I have to admit, I don't know what that is, but maybe it helps someone.

For now my temporary solution is to fix my `@types/react` dependency to version `16.7.13`.

**Comment by OliverJAsh** [comment]
@Manc 

> For now my temporary solution is to fix my `@types/react` dependency to version `16.7.13`.

We're using 16.7.8 and still have this problem, so I think you may be describing a different issue.

**Comment by OliverJAsh** [comment]
I created a failing test here: https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...OliverJAsh:oja/react-redux/connect-hoc#diff-d03480de14e7ab48b8f9da79ab9f4da8R1211

I'm working around the problem by patching the `InferableComponentEnhancerWithProps` type definition:

``` diff
 export type InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> =
-    <C extends ComponentType<Matching<TInjectedProps, GetProps<C>>>>(
+    <C extends ComponentType<GetProps<C>>>(
         component: C
-    ) => ConnectedComponentClass<C, Omit<GetProps<C>, keyof Shared<TInjectedProps, GetProps<C>>> & TNeedsProps>;
+    ) => ConnectedComponentClass<C, Omit<GetProps<C>, keyof TInjectedProps>>
```

This breaks some advanced usages, but none of them are ones we rely on. This will have to do until someone can help us identify a proper fix.

**Comment by VitamintK** [comment]
Having this issue as well

**Comment by svenlombaert** [comment]
Don't wanna be this guy, but having this issue as well.

Also, the workaround of @OliverJAsh works for me, haven't come across a case where it's not working in our projects. Any way to help/make sure this gets fixed?

**Comment by davnicwil** [comment]
After spending a couple of hours spinning on this, what solved it for me:

```jsx
const MyComponent = connect(
  null,
  // @ts-ignore
  mapDispatch,
)(Presentation)
```

Ugly, yes, but as a last resort it does just make it "go away". 

Since the error here doesn't really affect anything else, no harm done really.

**Comment by OliverJAsh** [comment]
> @apapirovski It would be great if you could share your understanding of how this code works and what you tried to achieve with it...

Given the author of these types (@apapirovski) doesn't seem to be responding, it seems the only way we're ever going to get this issue fixed is if someone either tries to understand the types (I've tried and failed), or someone completely rewrites them (this time with better documentation)—hopefully guided by the tests.

**Comment by OliverJAsh** [comment]
Partial response from @apapirovski https://twitter.com/apapirovski/status/1120703099822149638

**Comment by arnoldcui** [comment]
In my solution
```
Popout.tsx

interface IOwnProps {
  tabNodeId: string;
}

interface IDispatchProps {
  onPopout: () => void;
}

type IProps = IOwnProps & IDispatchProps;

const mapDispatchToProps = (dispatch: Dispatch, props: IOwnProps) => {
  return {
    onPopout: () => {
      console.log("test tabNodeId", props.tabNodeId);
    }
  };
};

const mapStateToProps = (state: StoreState) => {
  return {};
};

function HOC(Component: React.ComponentType) {
  const Mid = (props: IProps) => {
    return (
      <React.Fragment>
        <Component />
        <div>
          <h1>Now you too have the power to POP OUT</h1>
          <button onClick={props.onPopout}>{"Popout"}</button>
        </div>
      </React.Fragment>
    );
  };
  return connect(
    mapStateToProps,
    mapDispatchToProps
  )(Mid);
}

Counter.tsx

const CounterContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(Counter);

const CounterPopout = Popout(CounterContainer);

export {
  CounterPopout
};
```
I give the HOC function with a full props Component

**Comment by aczekajski** [comment]
@OliverJAsh Thank you for the workaround. It solved my problems with HOC but it created some new problems with components that were connected using OwnProps. I solved it by tweaking your solution a bit:
```diff
export type InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> =
-    <C extends ComponentType<Matching<TInjectedProps, GetProps<C>>>>(
+    <C extends ComponentType<GetProps<C>>>(
         component: C
-    ) => ConnectedComponentClass<C, Omit<GetProps<C>, keyof Shared<TInjectedProps, GetProps<C>>> & TNeedsProps>;
+    ) => ConnectedComponentClass<C, Omit<GetProps<C>, keyof TInjectedProps> & TNeedsProps>
```
As you can see I added the `TNeedProps` to the `P` parameter of `ConnectedComponentClass`.

**Comment by billyjacobs2014** [comment]
You can do the following with no side effects: Just replace the Matching<TInjectedProps, GetPropx<C>>>>> with any like so

`export type InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> =
    <C extends ComponentType<any>>(
        component: C
    ) => ConnectedComponentClass<C, Omit<GetProps<C>, keyof Shared<TInjectedProps, GetProps<C>>> & TNeedsProps>;`

**Comment by rodrigomf24** [comment]
I am not exactly sure if my issue is similar to others here but I've definitively saw a couple of cases where my code matched is some way, specially @19majkel94 scenario...

Anyways after wasting a few hours on what was going on I noticed that by removing an arrow function declaration inside the `mapDispatchToProps` I was no longer getting any TS errors.

Here is a sample of the code I was using to connect my component:

```javascript
const mapStateToProps = (state: ApplicationState) => {
  const {
    node: {
      anotherNode: {
        list,
        selectedItem,
        isDisabled,
      },
    },
  } = state;

  return {
    list,
    selectedItem,
    isDisabled,
  };
};

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      update,
      select,
      delete: (object: ObjectType) => null,
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SomeComponent);

```

The error that I was getting:

`Type '{ ... }' is not assignable to type 'Readonly<Pick<Props & ReduxProps, ...`

By removing this line `delete: (object: ObjectType) => null,`(the reason that line was in there in first place was because I was being lazy and didn't wanted to create the action, since I wanted to manually test other functionality first... 🤦‍♂ ) the error was gone 🎉 

I finally created the action like the rest and now it works, I hope this help other people to avoid spending so much time on the confusing or sometimes not so intuitive Typescript errors, but still I ❤️  Typescript.

**Comment by juicyarts** [comment]
Hey folks! I spent some time on this issue and wanted to share my observations. I used @OliverJAsh 's fix which helped a lot. This is what i ended up with:

```typescript
// these definitions come from other files but i added them to make it more reasonable 
interface ConnectorStateProps {
  user?: User;
}

interface ConnectorDispatchProps {
  loginUser: (credentials?: UserCredentials, signal?: AbortSignal) => Promise<SetUserAction>;
  setUser: (user: User) => SetUserAction;
}

function mapStateToProps(state: AppState): ConnectorStateProps {
  const { user } = state;
  return { user };
}

function mapDispatchToProps(dispatch: ThunkDispatch<AppState, {}, AnyAction>): ConnectorDispatchProps {
  return {
    loginUser: async (credentials, signal): Promise<void> => dispatch(UserActions.loginUser(credentials, signal)),
    setUser: (user): SetUserAction => dispatch(UserActions.setUser(user)),
  };
}

function withUser<P>(consumer: ComponentType<P>): ConnectedComponentClass<ComponentType<P>, Omit<P, keyof ConnectorStateProps | keyof ConnectorDispatchProps>> {
  const connector = connect<ConnectorStateProps, ConnectorDispatchProps, P, AppState>(mapStateToProps, mapDispatchToProps);
  const consume = connector(consumer);
  return consume;
}

export default withUser;
```

This is what VsCode displays for the return value of `const consume = connector(consumer);`:
![Screenshot 2019-07-01 at 16 30 52](https://user-images.githubusercontent.com/1132937/60444725-0ba7a980-9c1e-11e9-9cb4-f72eca7de5e6.png)

So the fix worked well here. Even using own props i did not have to extend the type definition for `InferableComponentEnhancerWithProps` like @aczekajski suggested.

There is little more to it though since i needed a return type for the `withUser` method. 
I had to provide `ConnectedComponentClass<ComponentType<P>, Omit<P, keyof ConnectorStateProps | keyof ConnectorDispatchProps>>` to make sure exclusion/inclusion is handled properly. 

In a Test i wrote one can observe that injected props are properly excluded while own props are passed through.

Heres the code for a TestComponent that i use to test the `withUser` function

```typescript
interface TestComponentProps extends UserConnectorStateProps, UserConnectorDispatchProps {
  foo: string;
}

function TestComponent({ user, foo }: TestComponentProps): FunctionComponentElement<TestComponentProps> {
  return (
    <div className="state">{ user && user.name }</div>
  );
}

const TestComponentWithUser = withUser<TestComponentProps>(TestComponent);
```

Hovering over the TestComponent in VsCode shows:
![Screenshot 2019-07-01 at 16 39 38](https://user-images.githubusercontent.com/1132937/60445655-ecaa1700-9c1f-11e9-83f6-7f0665e5f15a.png)

This works pretty well for me. I even tried out different overloads and types for the `connect` method and could not encounter issues so far.

So, im not really sure what the `ComponentType<Matching<TInjectedProps, GetProps<C>>>` is about and how it is intended to be used.. What i can say _from my point of view_ is the following:

For `InferableComponentEnhancerWithProps` it should be irrelevant what properties are passed to the component. Even if you change the definition of a property provided by `connect` in the components props, the definition in `connect` should overrule. Especially if the idea is to exclude the property from the props of the component when it's wrapped. 
So i guess this was implemented to make it easier to implement the return type of `InferableComponentEnhancerWithProps` but rather than just modifying the return type this is also setting an expectation for the requested properties type which, for me, looks like a bug..

**Comment by juicyarts** [comment]
> I created a failing test here: [master...OliverJAsh:oja/react-redux/connect-hocdiff-d03480de14e7ab48b8f9da79ab9f4da8R1211](https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...OliverJAsh:oja/react-redux/connect-hocdiff-d03480de14e7ab48b8f9da79ab9f4da8R1211)
> 
> I'm working around the problem by patching the `InferableComponentEnhancerWithProps` type definition:
> 
> ```diff
>  export type InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> =
> -    <C extends ComponentType<Matching<TInjectedProps, GetProps<C>>>>(
> +    <C extends ComponentType<GetProps<C>>>(
>          component: C
> -    ) => ConnectedComponentClass<C, Omit<GetProps<C>, keyof Shared<TInjectedProps, GetProps<C>>> & TNeedsProps>;
> +    ) => ConnectedComponentClass<C, Omit<GetProps<C>, keyof TInjectedProps>>
> ```
> 
> This breaks some advanced usages, but none of them are ones we rely on. This will have to do until someone can help us identify a proper fix.

Could you elaborate on the advanced use cases that break? Or does anyone else have an example of this fix breaking something to make sure we don't have any regressions when putting this into a PR ?

**Comment by OliverJAsh** [comment]
> Could you elaborate on the advanced use cases that break?

IIRC, some of the tests in DefinitelyTyped started to fail when I made this change. If you make the same change then run the tests, you should hopefully see what I meant.

**Comment by jralexandre** [comment]
I was having this error as well:
> Type '{}' is missing the following properties from type 'Readonly<Pick<Props, "ingredients" | "deleteIngredient">>': ingredients, deleteIngredient

I had my base component (not the connected) with these Props:
```
interface Props {
    deleteIngredient: (id: number) => void;
    ingredients: IngredientsType;
}
```

It was solved by changing it to:
`
type Props = StateProps & DispatchProps;
`

Where StateProps and DispatchProps are the same return types as my mapStateToProps and mapDispatchToProps functions. Somehow the signature was different and the compiler was complaining about both props not being present in the {} type (which is my empty OwnProps)

**Comment by eduard-malakhov** [comment]
I've also stumbled upon this issue and did some research. To me, the following two definitions seem internally inconsistent (cited from `@types/react-redux@7.1.8`).
```
export type InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> =
    <C extends ComponentType<Matching<TInjectedProps, GetProps<C>>>>(
        component: C
    ) => ConnectedComponent<C, Omit<GetProps<C>, keyof Shared<TInjectedProps, GetProps<C>>> & TNeedsProps>;
```
```
export type Matching<InjectedProps, DecorationTargetProps> = {
    [P in keyof DecorationTargetProps]: P extends keyof InjectedProps
        ? InjectedProps[P] extends DecorationTargetProps[P]
            ? DecorationTargetProps[P]
            : InjectedProps[P]
        : DecorationTargetProps[P];
};
```

The first definition implies that `C` is assignable to `ComponentType<Matching<TInjectedProps, GetProps<C>>>` which implies that `C` has the shape of `ComponentType<P>` where `P` is assignable to `Matching<TInjectedProps, P>`. Which in turn means that for all common properties between `P` and `Matching<TInjectedProps, P>` the type of the property in `P` is assignable to the type of that property in `Matching<TInjectedProps, P>`.

 From the definition for `Matching`, we see that the above is true for two of three branches of the double ternary operator—the first and the last one—because in those cases the type of a property in `Matching<TInjectedProps, P>` coincides with that in `P`. However, in the middle case, the type in `Matching<TInjectedProps, P>` will be inferred as `InjectedProps[P]` which as we know from the condition above is not assignable to the type of the same property in `P` (which exists due to condition is the line before that).

Hence, these two definitions imply that type of a property in `P` is assignable to `InjectedProps[P]` which in turn is **not** assignable to the type of that property in `P`. It looks like we've got a contradiction.

In my case, I've been able to eliminate the issue by making the following change to `@types/react-redux/index.d.ts` which as far as I can tell resembles the condition that was in effect before `Matching` was introduced.

```
  export type InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> =
-   <C extends ComponentType<Matching<TInjectedProps, GetProps<C>>>>(
+   <C extends ComponentType<Shared<TInjectedProps, GetProps<C>>>>(
          component: C
      ) => ConnectedComponent<C, Omit<GetProps<C>, keyof Shared<TInjectedProps, GetProps<C>>> & TNeedsProps>;
```

Does anyone know if any progress has been made in fixing this issue so far?

**Comment by JakeSidSmith** [comment]
We are now using our own adaptation of the connect types specifically for HOCs: https://github.com/dabapps/connect-hoc

Can be used alongside the existing inferable types.

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
